[Rui Ming] iP

ruiming97 opened the pull request:
## Neko

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Neko is here to help you manage your tasks, so you can focus on what matters. Whether it's tracking your to-do list or setting reminders for upcoming deadlines, Neko makes task management simple and efficient.

### Features
- Task Management
  - Easily add, view, and remove tasks using simple text commands.
- Upcoming Features
   - Deadline Management (coming soon)
       - Never miss a deadline again! Neko will help you track your important dates.
   - Reminders (coming soon)
       - Get notified when your tasks are due, so you can stay on top of everything.

## Installation
To start using Neko, follow these steps:

### Download Neko
Get the latest version from [here](https://github.com/ruiming97/ip/releases/tag/A-Jar).

### Run Neko
After downloading, simply double-click the application.
Add your tasks, and let Neko manage them for you!

## Why Use Neko?
- **Text-based**: No complicated UI to deal with. Just type your commands and get things done.
- **Fast**: Neko responds instantly, ensuring smooth task management.
- **Free**: Neko is completely free for all users.
Let Neko take care of your tasks, so you can focus on more important things!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename, Greet, and Exit

Commit:
Echo

Commit:
Add, List

Commit:
Mark as Done and improve code readability

Commit:
ToDos, Events, Deadlines

Commit:
Automated Text UI Testing

Commit:
Handle Errors

Commit:
Delete

Commit:
Add loadFIile and writeFile to save the data in data/neko.txt

Commit:
Merge branch 'branch-Level-7'

Commit:
Store date and time as a LocalDateTime for both Deadline and Event

Commit:
Merge branch 'branch-Level-8'

Commit:
Add removing the task from the file after it's deleted

Commit:
Make code more OOP by adding Parser, Storage, TaskList, Ui classes

Commit:
Merge branch 'more-oop'

Commit:
Divide classes into Neko and Neko.Task packages

Commit:
Merge branch 'add-gradle-support'

Commit:
change the name of package to start with small letter

Commit:
modify build.gradle to match with my setup

Commit:
modify the names of packages to start with small letter

Commit:
Add a few JUnit tests

Commit:
Merge branch 'junit-tests'

Commit:
Fix bug on JUnit tests

Commit:
Merge branch 'junit-tests'

Commit:
edit main class path and set jar name to neko.jar

Commit:
update text-ui testing

Commit:
Adds JavaDoc to some files

Commit:
Modify some lines to follow the coding standard

Commit:
Add find task with specified key word

Commit:
Merge branch 'java-doc'

* java-doc:
  Adds JavaDoc to some files

Commit:
Merge branch 'coding-standard'

* coding-standard:
  Modify some lines to follow the coding standard

# Conflicts:
#	src/main/java/neko/Neko.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find task with specified key word

# Conflicts:
#	src/main/java/neko/Neko.java
#	src/main/java/neko/task/TaskList.java

yxshrk commented on src/main/java/neko/Neko.java:
```diff
@@ -0,0 +1,149 @@
+import java.util.ArrayList;
+import java.io.IOException;
+
+/**
```
I like that you provided a well-detailed class-level JavaDoc. It provides good context. In the future, you may consider adding more details about edge cases or limitations to give future readers a better understanding of the design.

yxshrk commented on src/main/java/neko/Neko.java:
```diff
@@ -0,0 +1,149 @@
+        try {
+            tasks = new TaskList(storage.loadTasks());
+        } catch (FileNotFoundException e) {
+            ui.showMessage(e.getMessage());
```
You’re showing the error message through ui.showMessage(), which is helpful for the user. Have you considered adding a logging statement here as well, to track file-related issues for debugging purposes?

yxshrk commented on src/main/java/neko/Neko.java:
```diff
@@ -0,0 +1,149 @@
+        ui.showGreeting();
+
+        String input = ui.getUserCommand();
+        while (!input.equals(COMMAND_EXIT)) {
```
I like how you used a simple while condition to check for the exit command. Should we also handle the possibility of input being null to avoid potential NullPointerException?

yxshrk commented on src/main/java/neko/Parser.java:
```diff
@@ -0,0 +1,56 @@
+            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm");
+
+    public static Task parseTask(String str) {
+        char c = str.charAt(0);
```
Perhaps a more intuitive variable name for c here? taskType might make the code easier to read at a glance, especially for people unfamiliar with this part of the code. What do you think?

yxshrk commented on src/main/java/neko/Storage.java:
```diff
@@ -0,0 +1,82 @@
+
+public class Storage {
+
+    private static final String DIRECTORY_PATH = "./data";
```
I like how you made DIRECTORY_PATH a constant. Any particular reason why this is hardcoded and not configurable? Should we consider making it flexible for future changes?

yxshrk reviewed (commented):
Overall your code is well written. I just noticed a few minor coding standard violations and issues with the flexibility of the code.

Commit:
Do A-CheckStyle

Commit:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Do A-CheckStyle

Commit:
Do Level-10 GUI

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Do Level-10 GUI

Commit:
Do A-Varargs

Commit:
Merge branch 'branch-A-Varargs'

* branch-A-Varargs:
  Do A-Varargs

Commit:
Refactor TaskList class to add assertions and improve task handling.

Let's,
* add assertions to the TaskList constructor to ensure the task list is not null.
* add assertions in the `addTask` method to validate that the task being added is not null.
* update the `getTask` method with assertions to check the retrieved task is not null.
* refactore `markTask` and `unmarkTask` methods to improve readability:
  - add assertions to validate tasks are correctly marked as done or not done.
  - refactor loops to store the retrieved task in a variable before processing.
  - Simplify task marking logic by removing redundant code.
* sort task indices in descending order during deletion to avoid index errors when removing multiple tasks.

ruiming97 committed:
Merge pull request #1 from ruiming97/branch-A-Assertions

Refactor TaskList class to add assertions and improve task handling.

Commit:
Refactor a few classes to improve code quality

Let's,
 * add missing Javadoc
 * remove unnecessary imports from various classes
 * replace wildcard imports with explicit imports for clarity
 * refactor handleInput(String input) method to have better SLAP and avoid too long method
 * refactor createTask(String input) method to have better SLAP and avoid too long method
 * add a few private methods to achieve stronger SLAP

Commit:
Merge branch 'master' of https://github.com/ruiming97/ip

ruiming97 committed:
Merge pull request #3 from ruiming97/branch-A-CodeQuality

Refactor a few classes to improve code quality

Commit:
Merge branch 'master' of https://github.com/ruiming97/ip

Commit:
Add view schedule feature as an extension

Commit:
Merge branch 'branch-ViewSchedules'

Commit:
Improve the format of output for mark, unmark and delete

Commit:
Update the GUI to have automatic window sizing

Commit:
Fix adding task bug by updating the way how input is parsed

Commit:
Update README

Commit:
Merge branch 'branch-AUserGuide'

Commit:
Update README to include screenshot of the chatbot

Commit:
Try to fix that the table not being rendered correctly

Commit:
Third attempt to fix table not being rendered correctly

Commit:
Fix rendering bug

Commit:
Second attempt to fix rendering bug

Commit:
Add exit command to exit the chatbot

Commit:
Update readme to include ai acknowledgement
